Migration to DeclBaseName in diagnostic definitions

ahoppen opened the pull request:
These diagnostics are already called with DeclBaseName as their
parameters. Reflect this in the diagnostic's definition so that the
implicit conversion from DeclBaseName to Identifier is no longer needed.

It includes the changes from #9976

ahoppen commented:
Please test with following pull request:
apple/swift-lldb#206

@swift-ci Please test

jrose-apple commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -71,7 +71,7 @@ ERROR(could_not_find_pointer_pointee_property,none,
 ERROR(writeback_overlap_property,none,
       "inout writeback to computed property %0 occurs in multiple arguments to"
-      " call, introducing invalid aliasing", (Identifier))
+      " call, introducing invalid aliasing", (DeclBaseName))
```
This one's a property name, so it will never be a special name…but DeclBaseName is going to be more convenient from here on out. So the rule is "DeclBaseName unless you're actually talking about what the user wrote", I guess?

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2607,18 +2607,18 @@ NOTE(add_self_to_type,none,
 WARNING(warn_unqualified_access,none,
         "use of %0 treated as a reference to %1 in %2 %3",
-        (Identifier, DescriptiveDeclKind, DescriptiveDeclKind, DeclName))
+        (DeclBaseName, DescriptiveDeclKind, DescriptiveDeclKind, DeclName))
```
This one probably should be an Identifier, since it comes from something the user wrote. Yes, no?

jrose-apple reviewed (commented):
There are a lot of `Identifier` arguments left in DiagnosticsSema.def. What was the rule you used to decide which ones are operators?

ahoppen commented:
The basic rule was: Change everything from Identifier to DeclBaseName where the compiler complained about non-matching types. Should I have done/do more investigation on which changes are absolutely necessary?

jrose-apple commented:
I'm not sure that rule makes sense when your use of DeclBaseName vs. Identifier at the call site was mostly arbitrary. I'd rather be consistent, so that people who need to add new diagnostics know which one to use.

jrose-apple commented:
(I don't mean any offense by the call site being mostly arbitrary, just that it wasn't a focus.)

ahoppen commented:
OK, I'll look through them again and give you a ping when I'm done so you can have another look at the final result

ahoppen committed:
[Diagnostics] Adjustments for DeclBaseName

Adjust the definition of some diagnostics that are already called with
DeclBaseNames so that the implicit conversion from DeclBaseName to
Identifier is no longer needed.

Adjust the call side of diagnostics which don't have to deal with
special names to pass an Identifier to the diagnostic.

ahoppen commented:
@swift-ci please test

ahoppen commented:
@swift-ci please test

ahoppen commented:
@swift-ci Please test

ahoppen commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -2089,7 +2089,7 @@ ERROR(property_behavior_protocol_no_initStorage,none,
 ERROR(property_behavior_unknown_requirement,none,
       "property behavior protocol %0 has non-behavior requirement %1",
-      (Identifier, Identifier))
+      (Identifier, DeclBaseName))
```
@jckarter I changed this property behaviour related diagnostic I changed to `DeclBaseName`. Is that right or should it stay an `Identifier`?

jckarter replied:
It's probably best rendered as a `DeclName`, since a protocol requirement can have a compound name. Doesn't really matter since this isn't publicly exposed though.

ahoppen replied:
OK, I'll change it to `DeclName` then. Just thought it would be best to keep the hidden parts up-to-date as well.

jrose-apple reviewed (approved):
Looks good, with the one change we discussed (which can land after). Hooray for in-person reviews!

ahoppen commented:
@swift-ci Please test
